From 8b2c418af99e3f1e42884cd8fd7943445e6ac509 Mon Sep 17 00:00:00 2001
From: Brian Delwiche <delwiche@google.com>
Date: Mon, 22 Apr 2024 21:41:53 +0000
Subject: [PATCH] Add support for checking security downgrade

As a guard against the BLUFFS attack, we will need to check the security
parameters of incoming connections against cached values and disallow
connection if these parameters are downgraded or changed from their
cached values.

Future CLs will add checks during connection.  This CL adds the
functions that will be needed to perform those checks and the necessary
mocks.
Currently supported checks are : IO capabilities (must be an exact match),
Secure Connections capability (must not be a downgrade), and session key
length (must not be a downgrade).  Maximum session key length, which was
previously not cached, has been added to the device security manager
cache.

To QA: This CL is a logical no-op by itself.  Tests should be performed as described in ag/25815924 and ag/25815925/

Bug: 314331379
Test: m libbluetooth
Tag: #security
Ignore-AOSP-First: Security
Merged-In: I3d84b579c91c22165f3b7cfe31de2f6cde3f69fe
Merged-In: I3cd1db300be68d15cb09bdabea711199fcf748da
Merged-In: I972fd4a3a4d4566968d097df9f27396a821fb24f
Change-Id: I972fd4a3a4d4566968d097df9f27396a821fb24f
(cherry picked from commit 3cf3d9d98909787748e6135733b42be0c67e9333)
(cherry picked from commit c17811e6a2357eb34368a1a0a6ed5dec19d980ed)
---
 system/test/mock/mock_stack_btm_sec.cc | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/system/test/mock/mock_stack_btm_sec.cc b/system/test/mock/mock_stack_btm_sec.cc
index 7819435b89..6f7149ff04 100644
--- a/system/test/mock/mock_stack_btm_sec.cc
+++ b/system/test/mock/mock_stack_btm_sec.cc
@@ -121,6 +121,11 @@ bool btm_sec_is_session_key_size_downgrade(uint16_t hci_handle,
   inc_func_call_count(__func__);
   return false;
 }
+bool btm_sec_is_session_key_size_downgrade(uint16_t hci_handle,
+                                           uint8_t key_size) {
+  inc_func_call_count(__func__);
+  return false;
+}
 bool is_sec_state_equal(void* data, void* context) {
   inc_func_call_count(__func__);
   return false;
-- 
2.34.1

